mycpp/mylib_leaky.h

```diff
@@ -157,15 +157,42 @@ class Str : public gc_heap::Obj {
   }
   // s[begin:end]
   Str* slice(int begin, int end) {
+    begin = std::min(begin, len_);
+    end = std::min(end, len_);
+
+    assert(begin <= len_);
+    assert(end <= len_);
+
     if (begin < 0) {
       begin = len_ + begin;
     }
+
     if (end < 0) {
       end = len_ + end;
     }
+
+    begin = std::min(begin, len_);
+    end = std::min(end, len_);
+
+    begin = std::max(begin, 0);
+    end = std::max(end, 0);
+
+    assert(begin >= 0);
```

Comment:
I would write all of these

assert(0 <= begin && begin < len_);

Also it should be `< len_` not `<= len_`   (e.g. 0-based index )

And I think `min(begin, len_ - 1)`

Reply:
FWIW I actually strongly prefer 2 asserts with discrete conditions to one with a combined condition.  Makes it easier to see what the failure case was.

Re- the len_ conditionals; I'm actually quite sure the code is correct as-is.  There may be a different way of writing it, but I added more tests for the boundary conditions and it's working as it's written.  I did catch another corner case when writing those extra tests though. The one I caught was (begin > end).

Reply:
OK sorry I can see that end can be `len`.  BTW here is how Python does it, which is shorter, but also has some asymmetry that is a little hard to read:

https://github.com/oilshell/oil/blob/master/Python-2.7.13/Objects/stringobject.c#L1700